src/util/requests/impl/WsConnectionImpl.hpp

```diff
@@ -51,27 +57,46 @@ class WsConnectionImpl : public WsConnection {
     }
 
     std::expected<std::string, RequestError>
-    read(boost::asio::yield_context yield) override
+    read(boost::asio::yield_context yield, std::optional<std::chrono::steady_clock::duration> timeout = std::nullopt)
+        override
     {
         boost::beast::error_code errorCode;
         boost::beast::flat_buffer buffer;
 
-        ws_.async_read(buffer, yield[errorCode]);
+        auto operation = [&](auto&& token) { ws_.async_read(buffer, token); };
+        if (timeout) {
+            withTimeout(operation, yield[errorCode], *timeout);
+        } else {
+            operation(yield[errorCode]);
+        }
 
-        if (errorCode)
+        if (errorCode) {
+            errorCode = mapError(errorCode);
             return std::unexpected{RequestError{"Read error", errorCode}};
+        }
 
         return boost::beast::buffers_to_string(std::move(buffer).data());
     }
 
     std::optional<RequestError>
-    write(std::string const& message, boost::asio::yield_context yield) override
+    write(
+        std::string const& message,
+        boost::asio::yield_context yield,
+        std::optional<std::chrono::steady_clock::duration> timeout = std::nullopt
+    ) override
     {
         boost::beast::error_code errorCode;
-        ws_.async_write(boost::asio::buffer(message), yield[errorCode]);
-
-        if (errorCode)
+        auto operation = [&](auto&& token) { ws_.async_write(boost::asio::buffer(message), token); };
+        if (timeout) {
+            withTimeout(operation, yield[errorCode], *timeout);
+        } else {
+            operation(yield[errorCode]);
+        }
+
+        if (errorCode) {
+            errorCode = mapError(errorCode);
```

Comment:
Does the map error only need to happen when withTimeout? My concern is if it shadows any other error?

Reply:
No, it only to convert `operation_canceled` into `timed_out`.  You can see it at the line 138 in this file.